### DISCUSSION TITLE
Showing Dialogs Without Owner

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#8855] [Q&A]
I'm working on [HanumanInstitute.MvvmDialogs](https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/) to make Owner optional. This has a few use-cases:
- Showing the initial window
- Displaying message box on application start or end without a main UI

Implementing this in WPF, both MessageBox and framework dialogs (open/save dialog) auto-select the main window if available, and if not, still allow displaying it.

In Avalonia, however, open/save dialog REQUIRE owner to be set. As for messagebox, neither `MessageBox.Avalonia` nor `FluentAvalonia` works without setting the owner.

I'm trying to do a sample where you click the X, get a quit confirmation message, and then displays a messagebox on Closed event -- without a single line of code in the View. That last messagebox works in WPF but not Avalonia.

So... is there any particular reason why that wouldn't be possible in Avalonia to display such windows without setting the owner?

Adding to this, I'm aware that an unhandled exception will break Avalonia and prevent showing any further window. Would it be possible to re-initialize Avalonia just to display a final error message?

**Comment by timunie** [comment]
Are we talking about modal dialogs?

**Comment by kekekeks** [comment]
As for regular windows, just use `Show` instead of `ShowDialog`. You might need to subscribe to `Closed` event to convert it to a task.

As for file dialogs, there is no guarantee that the underlying platform supports showing them without a parent toplevel.

**Reply by timunie** [reply]
Can `Show` be a modal window? I.e.: Any other Window is blocked? Like when you show a `MessageBox` from `WPF`?

**Reply by kekekeks** [reply]
I don't think we can support application-wide modality due to restrictions of underlying platforms. E. g. with wayland you can't reliably activate a random window, so your user will be clicking on a disabled window while your "application modal" dialog isn't even visible on screen.

In general application-wide modality is just horrible UX artifact from the single-monitor era. Blocking a window on one monitor by a dialog from another window on a different monitor is a frustrating user experience.

**Reply by timunie** [reply]
I agree. Personally I like the dialogs that FluentAvalonia offers. An overlay over the content in semi transparent with message in the center of the window.

**Reply by mysteryx93** [reply]
I wrote on both [FluentAvalonia](https://github.com/amwx/FluentAvalonia/issues/189) and [MessageBox.Avalonia](https://github.com/AvaloniaCommunity/MessageBox.Avalonia/issues/97) to ask whether they could support it.

**Comment by mysteryx93** [comment]
OK I solved the problem. In new version of MvvmDialogs, if owner is null, I take the MainWindow if possible, or create a dummy parent window. Couldn't get the dummy window fully invisible though... for now it's leaving a gray square under the window.

**Comment by robloo** [comment]
I still think Avalonia should remove the parent window requirement: https://github.com/AvaloniaUI/Avalonia/issues/7934

**Comment by MathiasLui** [comment]
This was such a basic thing for me to expect coming from WPF, I'm surprised you now have to do `(App.Current.ApplicationLifetime as IClassicDesktopStyleApplicationLifetime).MainWindow` to show a basic modal window

**Reply by timunie** [reply]
Yeah that's the price for being cross platform. Mobile and wasm have no concept of Window, so no MainWindow available.

**Reply by MathiasLui** [reply]
Yeah that makes sense, wouldn't be the most unclean part of my code by far anyways

**Reply by Alshain01** [reply]
Alternatively, if you happen to be using dependency injection, you could create a Dialog service for injection that holds a reference to the Main Window.  Just be careful not to request it from the constructor of the Main window or anything constructed by the constructor of the Main Window (such as main view and main view model).

Also alternatively, and my personal reccomendation, learn to design UI without dialogs and message boxes.  Fly out menus and/or content presenters are much more portable between platforms and can serve the same purpose.  Dialogs are a 90's design that just won't seem to die... but should.

**Reply by mysteryx93** [reply]
MvvmDialogs handles this automatically; allows you to show without specifying parent

**Reply by maxkatz6** [reply]
> I'm surprised you now have to do

You don't have to. You SHOULDN'T use global statics. As it breaks whole concept of multiple windows and showing dialogs from the correct window.
Instead, your View should handle ViewModel requests to open dialogs (similarly to how ReactiveUI Interactions work) or use some kind of dependency injection. Or libraries that handle it automatically.

**Reply by robloo** [reply]
@maxkatz6 I think "SHOULDN'T" is much too strong of wording. In fact, Avalonia itself requires registering a main view and 99% of applications have a single main window/view from which all child windows or dialogs originate. You're speaking from a highly theoretical viewpoint and I don't think it relates well to real-world applications. In fact it makes real-world applications less efficient to write in cases like this.

Avalonia should allow showing a dialog with no owner/parent parameter. In that case it should fall back to the application's registered MainWindow if it exists.

**Reply by MathiasLui** [reply]
In my case it's strictly Desktop, also not using DI.
With "have to" i meant "have to, if I want to show a modal window", As it was going to be called from different views/VMs I wanted it to be shown in the middle, very simply. Having a flyout would also work but it would be much more fiddly for me as far as I can see.

I don't see it as 90s design but if you want to deploy your app x-platform, especially with mobile in mind, I gotta agree, but I also think a fallback to the registered MainWindow makes sense.

**Reply by rabbitism** [reply]
And from a **global static** point of view, is there really any difference between a global static and calling Application.Current.xxlifetime again and again? Application.Current is a global static, too.

**Reply by Alshain01** [reply]
It really isn't "fiddly" to make a side panel menu.  You just use a SplitView panel as your high level control.  Put your window content in the SplitView Content and when your button is clicked, set the SplitView.Pane content to whatever user control is appropriate for the pressed button and open it.  You will use the same SplitView control for all your side panel menus, just swap the Pane content.  It's your choice of course, Avalonia supports dialogs in the Desktop lifetime and you are free to use them, but the world has generally moved away from modal dialogs in UI design.

**Reply by timunie** [reply]
@MathiasLui check out the Awsome-Avalonia repo. For Example FluentAvalonia has some nice dialogs showing in the middle of the the window:

![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/e0626455-ee73-40eb-9870-30feb15e32e9)
